---
title: "*systemPipeR*'s New CWL Command-line Interface"
author: "Author: Daniela Cassol (danielac@ucr.edu) and Thomas Girke (thomas.girke@ucr.edu)"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`" 
output:
  BiocStyle::html_document:
    toc_float: true
    code_folding: show
  BiocStyle::pdf_document: default
package: systemPipeR
vignette: |
  %\VignetteEncoding{UTF-8}
  %\VignetteIndexEntry{systemPipeR's New CWL Command-line Interface}
  %\VignetteEngine{knitr::rmarkdown}
fontsize: 14pt
bibliography: bibtex.bib
---

```{css, echo=FALSE}
pre code {
white-space: pre !important;
overflow-x: scroll !important;
word-break: keep-all !important;
word-wrap: initial !important;
}
```

<!--
- Compile from command-line
Rscript -e "rmarkdown::render('systemPipeR.Rmd', c('BiocStyle::html_document'), clean=F); knitr::knit('systemPipeR.Rmd', tangle=TRUE)"; Rscript ../md2jekyll.R systemPipeR.knit.md 2; Rscript -e "rmarkdown::render('systemPipeR.Rmd', c('BiocStyle::pdf_document'))"
-->

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
options(width=60, max.print=1000)
knitr::opts_chunk$set(
    eval=as.logical(Sys.getenv("KNITR_EVAL", "TRUE")),
    cache=as.logical(Sys.getenv("KNITR_CACHE", "TRUE")), 
    tidy.opts=list(width.cutoff=60), tidy=TRUE)
```

```{r setup, echo=FALSE, messages=FALSE, warnings=FALSE}
suppressPackageStartupMessages({
    library(systemPipeR)
    library(BiocParallel)
    library(Biostrings)
    library(Rsamtools)
    library(GenomicRanges)
    library(ggplot2)
    library(GenomicAlignments)
    library(ShortRead)
    library(ape)
    library(batchtools)
})
```

**Note:** if you use _`systemPipeR`_ in published research, please cite:
Backman, T.W.H and Girke, T. (2016). *systemPipeR*: NGS Workflow and Report Generation Environment. *BMC Bioinformatics*, 17: 388. [10.1186/s12859-016-1241-0](https://doi.org/10.1186/s12859-016-1241-0).

# *systemPipeR*'s New CWL Command-line Interface

Computational workflows are becoming increasingly important for advanced scientific research,
mainly because of the amount of data output by next-generation sequencing technologies. 
Workflows typically are composed of multiple software tools, or pipelines, each with a specific
set of parameters, a different configuration of the input data and also the output results. 
Reproducibility and scalable are the main concern when a research workflow is designed. 
Although several tools for managing and executing workflow systems are available, they are designed 
to offer a specific set of functionalities. To address this need, we have developed a new tool 
(`sysagrs2` - S4 Class) to create and run workflows with `systempipeR` package, integrating the Common Workflow Language (CWL),
which provide a standard for describing analysis workflows in a generic and reproducible manner.


## CWL Introduction 
<!--
TODO
-->

## Structure of _`new param`_ files and _`SYSargs2`_ container
The _`.cwl`_ files defines the parameters of a chosen command-line software. The following shows the format of a sample _`hisat2-mapping-se.cwl`_ file provided by this package. 

```{r cwl_structure, eval=TRUE}
library(systemPipeR)
targets <- system.file("extdata", "targets.txt", package="systemPipeR")
dir_path <- system.file("extdata/cwl/hisat2-se", package="systemPipeR")
WF <- loadWorkflow(targets=targets, wf_file="hisat2-mapping-se.cwl",
                   input_file="hisat2-mapping-se.yml",
                   dir_path=dir_path)

WF <- renderWF(WF, inputvars=c(FileName="_FASTQ_PATH_", SampleName="_SampleName_"))
WF
clt(WF)
```

*SYSargs2 class* stores all the information and instructions needed for processing 
a set of input files with a specific command-line or a series of command-line within 
a workflow. The *SYSargs2* S4 class object is created from the *loadWorkflow* 
and *renderWF* function, which populates all the command-line for each sample in 
each step of the particular workflow. Each sample level input/outfile operation uses its 
own *SYSargs2* instance. The output of *SYSargs2* define all the expected 
output files for each step in the workflow, which usually it is the sample input for the
next step in an *SYSargs2* instance. Between different instances, this connectivity
is established by writing the subsetting output with the *writeTargetsout* function
to a new targets file that serves as input to the next *loadWorkflow* and *renderWF* 
call. By chaining several *SYSargs2* steps together one can construct complex workflows 
involving many sample-level input/output file operations with any combination of command-line
or R-based software. 

Several accessor methods are available that are named after the slot names of the _`SYSargs2`_ object. 

```{r names_WF, eval=TRUE}
names(WF)
```

Of particular interest is the _`cmdlist()`_ method. It constructs the system
commands for running command-lined software as specified by a given _`.cwl`_
file combined with the paths to the input samples (_e.g._ FASTQ files) provided
by a _`targets`_ file. The example below shows the _`cmdlist()`_ output for
running HISAT2 on the first SE read sample. Evaluating the output of
_`cmdlist()`_ can be very helpful for designing and debugging _`.cwl`_ files
of new command-line software or changing the parameter settings of existing
ones.  

```{r cmdlist, eval=TRUE}
cmdlist(WF)[1]
modules(WF)
targets(WF)[1]
targets.as.df(targets(WF))
output(WF)
cwlfiles(WF)
inputvars(WF)
infile1(WF)[1:4]
```

# Showcase Workflow with `HISAT2`

## Read mapping with `HISAT2`

The NGS reads of this project will be aligned against the reference
genome sequence using `Hisat2` [@Kim2015-ve]. The parameter 
settings of the aligner are defined in the `workflow_hisat2-se.cwl` and `workflow_hisat2-se.yml` files.

```{r hisat_alignment2, eval=FALSE}
#######################################
## Example: hisat2 WF with samtools ##
######################################
targets <- system.file("extdata", "targets.txt", package="systemPipeR")
dir_path <- system.file("extdata/cwl/workflow-hisat2-se", package="systemPipeR")
WF <- loadWorkflow(targets=targets, wf_file="workflow_hisat2-se.cwl",
                   input_file="workflow_hisat2-se.yml",
                   dir_path=dir_path)
WF <- renderWF(WF, inputvars=c(FileName="_FASTQ_PATH_", SampleName="_SampleName_"))
WF

## Paired-End HISAT2 only
targetsPE <- system.file("extdata", "targets.txt", package="systemPipeR")
dir_path <- system.file("extdata/cwl/hisat2-pe", package="systemPipeR")
WF <- loadWorkflow(targets=targetsPE, wf_file="hisat2-mapping-pe.cwl",
                   input_file="hisat2-mapping-pe.yml",
                   dir_path=dir_path)
WF <- renderWF(WF, inputvars=c(FileName1="_FASTQ_PATH1_", FileName2="_FASTQ_PATH2_", SampleName="_SampleName_"))
WF

## Paired-End HISAT2 WF
dir_path <- system.file("extdata/cwl/workflow-hisat2-pe", package="systemPipeR")
WF <- loadWorkflow(targets=targetsPE, wf_file="workflow_hisat2-pe.cwl",
                   input_file="workflow_hisat2-pe.yml",
                   dir_path=dir_path)
WF <- renderWF(WF, inputvars=c(FileName1="_FASTQ_PATH1_", FileName2="_FASTQ_PATH2_", SampleName="_SampleName_"))
WF
```

Subsetting _`SYSargs2`_ class slots for each workflow step.

```{r subset, eval=FALSE}
## Testing subset_wf function
subsetWF(WF, slot="input", subset='FileName')
subsetWF(WF, slot="output", subset=2)
subsetWF(WF, slot="step", subset=1) ## subset all the HISAT2 commandline 
subsetWF(WF, slot="output", subset="samtools-index.cwl")
subsetWF(WF, slot="output", subset=1, delete=TRUE) ##DELETE
```

Execute _`SYSargs2`_ on a single machine without submitting to a queuing system of a compute cluster. This way the input FASTQ files will be processed sequentially.

```{r runCommandline_WF, eval=FALSE}
## runCommandLine
library(systemPipeR)
runCommandline(WF) ## creates the files in the ./results folder
runCommandline(WF, dir=TRUE) ## creates the files in the ./results/workflowName/Samplename folder
runCommandline(WF, dir=TRUE, make_bam = TRUE) ##if it uses the workflow with samtools, should not uses make_bam=TRUE
```

Check and update the output location if necessary.

```{r output, eval=FALSE}
WF <- output_update(WF, dir=TRUE) ## Updates the output(WF) to the right location in the subfolders
WF <- output_update(WF, dir=TRUE, replace = ".bam") ## Updates the output(WF) to the right location in the subfolders
output(WF)
## Add to runCommandline
```

Check whether all BAM files have been created.

```{r WF_track, eval=FALSE}
WF_track <- run_track(WF_ls = c(WF))
names(WF_track)
WF_steps(WF_track)
track(WF_track)
summaryWF(WF_track)
```

Parallelization of read/alignment stats via scheduler (_e.g._ Slurm) across several compute nodes.

```{r clusterRun, eval=FALSE}
## clusterRun
library(batchtools)
resources <- list(walltime=120, ntasks=1, ncpus=4, memory=1024)
reg <- clusterRun2(WF, FUN=runCommandline2, conffile = ".batchtools.conf.R", template = "batchtools.slurm.tmpl", Njobs=4, runid="01", resourceList=resources)
getStatus(reg=reg)

WF <- output_update(WF, dir=TRUE) ## Updates the output(WF) to the right location in the subfolders
output(WF)
```

### Read and alignment stats

The following provides an overview of the number of reads in each sample
and how many of them aligned to the reference.

```{r align_stats, eval=FALSE}
read_statsDF <- alignStats(args=WF) 
write.table(read_statsDF, "results/alignStatsWF.xls", row.names=FALSE, quote=FALSE, sep="\t")
read_statsDF
```

### Write _`new targets files`_

To establish the connectivity between different instances, it is possible by writing the subsetting output with the _`writeTargetsout`_ function to a new targets file that serves as input to the next _`loadWorkflow`_ and _`renderWF`_ call.

```{r writeTargetsout, eval=FALSE}
names(WF$clt)
writeTargetsout(x=WF, file="default", step=1)
```


## Building a custom workflow with CWL

 TODO: Describe the rules and standards for _`systemPipeR`_ and how to create a new template using the _`create.clt`_.

<!--
- provide cwltool design tool
-->

## Showcase small RNA-Seq workflow



## Gunzip example
```{r gunzip, eval=FALSE}
WF <- loadWorkflow(targets="./targets.txt", wf_file="gunzip.cwl",
                   input_file="gunzip.yml",
                   dir_path="param/cwl/gunzip/")
WF <- renderWF(WF, inputvars=c(FileName="_FASTQ_PATH_", SampleName="_SampleName_"))
WF

WF <- loadWorkflow(targets="./targets.txt", wf_file="fastx_clipper.cwl",
                   input_file="fastx_clipper.yml",
                   dir_path="param/cwl/fastx_toolkit/")
WF <- renderWF(WF, inputvars=c(FileName="_FASTQ_PATH_", SampleName="_SampleName_"))
WF

WF <- loadWorkflow(targets="./targets.txt", wf_file="workflow-fastx_clipper.cwl",
                   input_file="workflow-fastx_clipper.yml",
                   dir_path="param/cwl/workflow-fastx_clipper/")
WF <- renderWF(WF, inputvars=c(FileName="_FASTQ_PATH_", SampleName="_SampleName_"))
WF
```



# Version Information

```{r sessionInfo}
sessionInfo()
```

# Funding

This project is funded by NSF award [ABI-1661152](https://www.nsf.gov/awardsearch/showAward?AWD_ID=1661152).

# References
